fdbserver/ClusterController.actor.cpp

```diff
@@ -1214,40 +1214,44 @@ class ClusterControllerData {
-						    conf, required, desired, policy, testUsed, checkStable, dcIds, exclusionWorkerIds);
-						RoleFitness testFitness(testWorkers, ProcessClass::TLog, testUsed);
-						RoleFitness fitness(workers, ProcessClass::TLog, id_used);
-
-						std::map<Optional<Standalone<StringRef>>, int> field_count;
-						std::set<Optional<Standalone<StringRef>>> zones;
-						for (auto& worker : testWorkers) {
-							if (!zones.count(worker.interf.locality.zoneId())) {
-								field_count[worker.interf.locality.get(pa1->attributeKey())]++;
-								zones.insert(worker.interf.locality.zoneId());
+						try {
+							auto testWorkers = getWorkersForTlogsBackup(
+							    conf, required, desired, policy, testUsed, checkStable, dcIds, exclusionWorkerIds);
+							RoleFitness testFitness(testWorkers, ProcessClass::TLog, testUsed);
+							RoleFitness fitness(workers, ProcessClass::TLog, id_used);
+
+							std::map<Optional<Standalone<StringRef>>, int> field_count;
+							std::set<Optional<Standalone<StringRef>>> zones;
+							for (auto& worker : testWorkers) {
+								if (!zones.count(worker.interf.locality.zoneId())) {
+									field_count[worker.interf.locality.get(pa1->attributeKey())]++;
+									zones.insert(worker.interf.locality.zoneId());
+								}
 							}
-						}
-						// backup recruitment is not required to use degraded processes that have better fitness
-						// so we cannot compare degraded between the two methods
-						testFitness.degraded = fitness.degraded;
-
-						int minField = 100;
+							// backup recruitment is not required to use degraded processes that have better fitness
+							// so we cannot compare degraded between the two methods
+							testFitness.degraded = fitness.degraded;
 
-						for (auto& f : field_count) {
-							minField = std::min(minField, f.second);
-						}
+							int minField = 100;
 
-						if (fitness > testFitness && minField > 1) {
-							for (auto& w : testWorkers) {
-								TraceEvent("TestTLogs").detail("Interf", w.interf.address());
+							for (auto& f : field_count) {
+								minField = std::min(minField, f.second);
 							}
-							for (auto& w : workers) {
-								TraceEvent("RealTLogs").detail("Interf", w.interf.address());
+
+							if (fitness > testFitness && minField > 1) {
+								for (auto& w : testWorkers) {
+									TraceEvent("TestTLogs").detail("Interf", w.interf.address());
+								}
+								for (auto& w : workers) {
+									TraceEvent("RealTLogs").detail("Interf", w.interf.address());
+								}
+								TraceEvent("FitnessCompare")
+								    .detail("TestF", testFitness.toString())
+								    .detail("RealF", fitness.toString());
+								ASSERT(false);
 							}
-							TraceEvent("FitnessCompare")
-							    .detail("TestF", testFitness.toString())
-							    .detail("RealF", fitness.toString());
-							ASSERT(false);
+						} catch (Error& e) {
+							ASSERT(false); // Simulation only validation should not throw errors
```

Comment:
```suggestion
							// Simulation only validation should not throw errors
							ASSERT(e.code() == error_code_actor_cancelled || e.code() == invalid_error_code); 
```

`error_code_actor_cancelled` is probably not necessary as this code block doesn't wait, but it would probably be good to have a good template here... Checking for `invalid_error()` though probably makes the errors slightly more readable. Otherwise an assertion failure will cause another assertion failure